crates/ra_hir_ty/src/traits/chalk.rs

```diff
@@ -130,11 +131,34 @@ impl<'a> chalk_solve::RustIrDatabase<Interner> for ChalkContext<'a> {
         self.db.program_clauses_for_chalk_env(self.krate, environment.clone())
     }
 
-    fn opaque_ty_data(
-        &self,
-        _id: chalk_ir::OpaqueTyId<Interner>,
-    ) -> Arc<rust_ir::OpaqueTyDatum<Interner>> {
-        unimplemented!()
+    fn opaque_ty_data(&self, id: chalk_ir::OpaqueTyId<Interner>) -> Arc<OpaqueTyDatum> {
+        let interned_id = crate::db::InternedOpaqueTyId::from(id);
+        let full_id = self.db.lookup_intern_impl_trait_id(interned_id);
+        let (func, idx) = match full_id {
+            crate::OpaqueTyId::ReturnTypeImplTrait(func, idx) => (func, idx),
+        };
+        let datas =
+            self.db.return_type_impl_traits(func).expect("impl trait id without impl traits");
+        let data = &datas.value.impl_traits[idx as usize];
+        let bound = OpaqueTyDatumBound {
+            bounds: make_binders(
+                data.bounds
+                    .value
+                    .iter()
+                    .cloned()
+                    .filter(|b| !b.is_error())
+                    .map(|b| b.to_chalk(self.db))
+                    .collect(),
+                1,
+            ),
+        };
+        let num_vars = datas.num_binders;
+        Arc::new(OpaqueTyDatum { opaque_ty_id: id, bound: make_binders(bound, num_vars) })
+    }
+
+    fn hidden_opaque_type(&self, _id: chalk_ir::OpaqueTyId<Interner>) -> chalk_ir::Ty<Interner> {
+        // FIXME: actually provide the hidden type; it is relevant for auto traits
```

Comment:
This is a fun one :)

Off the top of my head, I am not sure if we should make this lazy or not. I guess we should try, but I won't be surprised if most of the time something somewhere asks a question about `: Sync`.

Reply:
It actually needs to be lazy, in that Chalk needs to only ask for it when it needs to prove an auto trait goal. Otherwise valid recursive calls will result in cycle errors. E.g. compare [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=285acda9b75ce760bbe6196130bb95b2) with uncommenting the commented-out line...

Not sure if you mean something else by lazy :slightly_smiling_face:

Reply:
Hm, interesting example! Yeah, I guess we than don't have a choice here :-)